src/riak_client.erl

```diff
@@ -398,20 +399,45 @@ delete(Bucket,Key,Options,Timeout,{?MODULE, [Node, _ClientId]}=THIS) when is_lis
         true ->
             consistent_delete(Bucket, Key, Options, Timeout, THIS);
         false ->
-            normal_delete(Bucket, Key, Options, Timeout, THIS);
+            maybe_normal_delete(Bucket, Key, Options, Timeout, THIS);
         {error,_}=Err ->
             Err
     end;
 delete(Bucket,Key,RW,Timeout,{?MODULE, [_Node, _ClientId]}=THIS) ->
     delete(Bucket,Key,[{rw, RW}], Timeout, THIS).
 
+%% If this is a write-once bucket, delete without read-before-delete
+%% (write_once_delete).  Else do a normal delete.
+
+maybe_normal_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]}) ->
+    case write_once(Node, Bucket) of
+        true ->
+	    write_once_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]});
+        false ->
+	    normal_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]});
+        {error,_}=Err ->
+            Err
+    end.
+
+%% This version is exported, since it is called directly from
+%% riak_kv_ts_api:delete_data/4
+
+write_once_delete(Bucket, Key, Options, {?MODULE, [Node, ClientId]}) ->
+    write_once_delete(Bucket, Key, Options, ?DEFAULT_TIMEOUT, {?MODULE, [Node, ClientId]}).
+
+%% Write-once get path ignores vclocks.  We pass in an empty vclock
+%% (vclock:fresh()) so that no read-before-delete is triggered in
+%% riak_kv_delete:delete/8
+
+write_once_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]}) ->
+    normal_delete_vclock(Bucket, Key, vclock:fresh(), Options, Timeout, {?MODULE, [Node, ClientId]}).
```

Comment:
I'm a bit concerned about `vclock:fresh()` here - given other concurrent writes using the logic in [`w1c_vclock`](https://github.com/basho/riak_kv/blob/f8fbc77af4e27395124aaae07715de4b74a90855/src/riak_kv_w1c_worker.erl#L217), and the new "take the last timestamp" logic when resolving values, I think we may end up with deletes _always losing_ against a concurrent write (and remember concurrent doesn't mean "at the same time" - the delete could be later, but end up being overwritten by a node that comes back online and ends up repairing the value).